httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveDataProducer.java

```diff
@@ -0,0 +1,175 @@
+ */
+@Contract(threading = ThreadingBehavior.SAFE)
+final class ReactiveDataProducer implements AsyncDataProducer, Subscriber<ByteBuffer> {
+
+    private static final int BUFFER_WINDOW_SIZE = 5;
+
+    private final AtomicReference<DataStreamChannel> requestChannel = new AtomicReference<>();
+    private final AtomicReference<Throwable> exception = new AtomicReference<>(null);
+    private final AtomicBoolean complete = new AtomicBoolean(false);
+    private final Publisher<ByteBuffer> publisher;
+    private final AtomicReference<Subscription> subscription = new AtomicReference<>(null);
+    private final ArrayDeque<ByteBuffer> buffers = new ArrayDeque<>(); // This field requires synchronization
+
+    public ReactiveDataProducer(final Publisher<ByteBuffer> publisher) {
+        this.publisher = Args.notNull(publisher, "publisher");
+    }
+
+    @Override
+    public void onSubscribe(final Subscription subscription) {
+        if (this.subscription.getAndSet(subscription) != null) {
+            throw new IllegalStateException("Already subscribed");
+        }
+
+        subscription.request(BUFFER_WINDOW_SIZE);
+    }
+
+    @Override
+    public void onNext(final ByteBuffer byteBuffer) {
+        final byte[] copy = new byte[byteBuffer.remaining()];
+        byteBuffer.get(copy);
+        synchronized (buffers) {
+            buffers.add(ByteBuffer.wrap(copy));
+        }
+        signalReadiness();
+    }
+
+    @Override
+    public void onError(final Throwable throwable) {
+        subscription.set(null);
+        exception.set(throwable);
+        signalReadiness();
+    }
+
+    @Override
+    public void onComplete() {
+        subscription.set(null);
+        complete.set(true);
+        signalReadiness();
+    }
+
+    private void signalReadiness() {
+        final DataStreamChannel channel = requestChannel.get();
+        if (channel == null) {
+            throw new IllegalStateException("Output channel is not set");
+        }
+        channel.requestOutput();
+    }
+
+    @Override
+    public int available() {
+        if (exception.get() != null || complete.get()) {
+            return 1;
+        } else {
+            synchronized (buffers) {
+                int sum = 0;
+                for (final ByteBuffer buffer : buffers) {
+                    sum += buffer.remaining();
+                }
+                return sum;
+            }
+        }
+    }
+
+    @Override
+    public void produce(final DataStreamChannel channel) throws IOException {
+        if (requestChannel.get() == null) {
+            requestChannel.set(channel);
+            publisher.subscribe(this);
+        }
+
+        final Throwable t = exception.get();
+        final Subscription s = subscription.get();
+        int buffersToReplenish = 0;
+        try {
+            synchronized (buffers) {
+                if (t != null) {
+                    // TODO: We need a reliable way to send RST_STREAM (without subsequent GOAWAY) on HTTP/2.
```

Comment:
@rschmitt By the way, why not using a subclass of `H2StreamResetException` here?

Reply:
That's an interesting option, but what happens if I throw it on HTTP/1.1?

Reply:
@rschmitt Nothing spectacular. I'll likely be handled like any other I/O error. I am still open to using a special annotation for exceptions, though, if you like that approach better.

Reply:
I'll experiment with `H2StreamResetException`. The suggestion to use annotations is creative, but it may be overkill for what I need, which is probably not much.

Reply:
`H2StreamResetException` works perfectly, except for the additional dependency edge from `:httpcore5-reactive` to `:httpcore5-h2`, which I think is acceptable.

Reply:
@rschmitt I'll refactor that bit once the change-set gets merged.